analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs

```diff
@@ -27,122 +28,110 @@
-                                {
-                                    continue;
-                                }
-
-                                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, invalidCreation.Syntax.GetLocation(),
-                                    typeName));
-                            }
-                        });
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(ProcessObjectCreations(symbolsWhereTypeIsCreated),
+                                                                                   SyntaxKind.ObjectCreationExpression,
+                                                                                   SyntaxKindEx.ImplicitObjectCreationExpression);
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(ProcessSimpleAssignments(symbolsWhereLocaleIsSet), SyntaxKind.SimpleAssignmentExpression);
+                    compilationStartContext.RegisterCompilationEndAction(ProcessCollectedSymbols(symbolsWhereTypeIsCreated, symbolsWhereLocaleIsSet));
                 });
-        }
 
-        private static SyntaxNode GetAssignmentTargetVariable(ObjectCreationExpressionSyntax objectCreation)
-        {
-            var parent = objectCreation.GetFirstNonParenthesizedParent();
+        private static Action<SyntaxNodeAnalysisContext> ProcessObjectCreations(ConcurrentBag<SyntaxNodeWithSymbol<SyntaxNode, ISymbol>> symbolsWhereTypeIsCreated) =>
+            c =>
+            {
+                if (!(GetSymbolFromConstructorInvocation(c.Node, c.SemanticModel) is ITypeSymbol objectType) || !objectType.IsAny(CheckedTypes))
+                {
+                    return;
+                }
 
-            if (parent is AssignmentExpressionSyntax assignment)
+                var variableSyntax = GetAssignmentTargetVariable(c.Node);
+                if (variableSyntax == null)
+                {
+                    return;
+                }
+
+                var variableSymbol = variableSyntax is IdentifierNameSyntax
+                    ? c.SemanticModel.GetSymbolInfo(variableSyntax).Symbol
+                    : c.SemanticModel.GetDeclaredSymbol(variableSyntax);
+                if (variableSymbol != null)
+                {
+                    symbolsWhereTypeIsCreated.Add(variableSymbol.ToSymbolWithSyntax(c.Node));
+                }
+            };
+
+        private static Action<SyntaxNodeAnalysisContext> ProcessSimpleAssignments(ConcurrentBag<ISymbol> symbolsWhereLocaleIsSet) =>
+            c =>
             {
-                return assignment.Left;
-            }
+                var assignmentExpression = (AssignmentExpressionSyntax)c.Node;
+                var propertySymbol = GetPropertySymbol(assignmentExpression, c.SemanticModel);
 
-            if (parent is EqualsValueClauseSyntax equalsClause &&
-                equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration)
+                if (propertySymbol != null
```

Comment:
I'd remove the standalone `var propertySymbol` line
```suggestion
                if (GetPropertySymbol(assignmentExpression, c.SemanticModel); is {} propertySymbol
```